src/mpl/src/sock/mpl_sockaddr.c

```diff
@@ -0,0 +1,316 @@
+/*
+*  (C) 2018 by Argonne National Laboratory.
+*      See COPYRIGHT in top-level directory.
+*/
+
+/** Rationale:
+ *    MPL wrap for handling IPv4 and IPv6.
+ *
+ *    Applications: pm, pmi, ch3.
+ *    ch4 supports tcp sockets indirectly through ucx and ofi.
+ */
+
+/** Design considerations:
+ *    Either IPv4 or IPv6, globally set as defalt or with command line option, to
+ *    simplify logic.
+ *    TCP only, no UDP or unix domain sockets.
+ *
+ *    Application use struct sockaddr_storage exclusively.
+ *            MPL_get_sockaddr for hostname
+ *            MPL_get_sockaddr_iface for network interface
+ *            MPL_get_sockaddr_direct for listening socket on ANY or LOOPBACK
+ *
+ *    Simplified MPL_connect and MPL_listen interface.
+ *            Both have a port parameter.
+ *            MPL_listen combines bind with listen.
+ */
+
+/** Portability:
+ *    struct sockaddr_storage:
+ *            In case this struct is not available (in sys/socket.h), it can be
+ *        circumvented by declare following (in mpl_sockaddr.h):
+ *                    struct sockaddr_storage {
+ *                            unsigend short ss_family;
+ *                            char padding[126];
+ *                    };
+ *            Only the ss_family field is directly accessed. All the other fields are
+ *        always accessed by casting to either struct sockaddr_in or struct
+ *        sockaddr_in6.
+ *
+ *    The implementation uses getaddrinfo and getifaddrs. The former, as with
+ *    sockaddr_storage and sockaddr_in6, are documented in RFC 2553, 1999, and are
+ *    expected to be supported on most supported platforms. getifaddrs is not in
+ *    POSIX.1, but it is present on Linux since glibc 2.3.3, and available on BSD
+ *    systems even earlier.
+ */
+
+#include "mplconfig.h"
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netdb.h>
+#include <netinet/in.h>
+#include <string.h>
+#include <ifaddrs.h>
+#include <errno.h>
+#include <stdio.h>
+
+#include "mpl_sockaddr.h"
+
+static int is_localhost(struct sockaddr *p_addr);
+
+static int af_type = AF_INET;
+static int _use_loopback = 0;
+static int _max_conn = SOMAXCONN;
+
+int MPL_get_sockaddr(const char *s_hostname, struct sockaddr_storage *p_addr)
+{
+    struct addrinfo ai_hint;
+    struct addrinfo *ai_list;
+    int ret;
+
+    memset(p_addr, 0, sizeof(*p_addr));
+    memset(&ai_hint, 0, sizeof(ai_hint));
+    ai_hint.ai_family = af_type;
+    ai_hint.ai_socktype = SOCK_STREAM;
+    ai_hint.ai_protocol = IPPROTO_TCP;
+    ai_hint.ai_flags = AI_ADDRCONFIG | AI_V4MAPPED;
+    ret = getaddrinfo(s_hostname, NULL, &ai_hint, &ai_list);
+    if (ret) {
+        return ret;
+    }
+    ret = 0;
+    if (af_type == AF_INET) {
+        memcpy(p_addr, ai_list->ai_addr, sizeof(struct sockaddr_in));
+    } else if (af_type == AF_INET6) {
+        memcpy(p_addr, ai_list->ai_addr, sizeof(struct sockaddr_in6));
+    } else {
+        ret = 1;
```

Comment:
I prefer to abort here since `af_type` is having an unexpected value here. Something must be really wrong if we hit this branch.

Reply:
Remember this is down in MPL, not MPI land. It should be the responsibility of MPICH to abort where there is a failure.

Reply:
Yanfei is referring to the case that should not happen or is simply error.  I  am replacing these cases with simply assert(0).